### DISCUSSION TITLE
Event sourced process managers

### DISCUSSION THREAD

**Discussion by slashdotdash** [commanded/commanded#582] [Ideas]
Process managers currently persist their state after each handled event. They also use a single all stream subscription to route events to an instance and for the instance to handle the event. 

The next release of [EventStore has support for linking events](https://github.com/commanded/eventstore/pull/103). This feature makes it possible to use event sourcing to build a process manager instance's state. Events can be linked from the source stream to a stream dedicated to the process manager instance. To rebuild its state will require fetching and applying the events, exactly as aggregates work. 

Using a stream of linked events will also allow individual process manager instances to have their own subscription to their own stream of events. This will allow instances to process events concurrently, rather than being blocked by each other. The process router will still use an all stream subscription, but it will now just be linking events to the target instance's event stream. The instance's own subscription will trigger its event handling and command dispatch.

Due to the potentially large number of process manager instances running they should be automatically shutdown after an inactivity timeout period (#148).

The benefit of this approach is that event sourcing allows state to be modified by changing the process manager behaviour and rebuilding. Whereas currently the state cannot be amended from historical events.

It should still be possible to snapshot process manager instance state using the same snapshot behaviour as aggregates.

**Comment by dvic** [comment]
@slashdotdash there is one thing that's unclear: where does the linking happen? Do you have to do that manually? Or is this configured as part of the eventstore adapter to do this automatically for you? how can we otherwise reliably link events (currently)?

**Comment by lachiemurray** [comment]
> @slashdotdash there is one thing that's unclear: where does the linking happen? Do you have to do that manually? Or is this configured as part of the eventstore adapter to do this automatically for you? how can we otherwise reliably link events (currently)?

Also interested in this. I could see it being done in the eventstore adapter on event creation, but if you later changed you mind about what events should be part of the stream that would presumably have to be done manually as a one off. Curious to know if you explored any possibilities @dvic ?
